server/src/main/java/org/opensearch/telemetry/tracing/handler/TraceableTransportResponseHandler.java

```diff
@@ -0,0 +1,87 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.telemetry.tracing.handler;
+
+import org.opensearch.common.util.FeatureFlags;
+import org.opensearch.core.common.io.stream.StreamInput;
+import org.opensearch.core.transport.TransportResponse;
+import org.opensearch.telemetry.tracing.Span;
+import org.opensearch.transport.TransportException;
+import org.opensearch.transport.TransportResponseHandler;
+
+import java.io.IOException;
+import java.util.Objects;
+
+/**
+ * Tracer wrapped {@link TransportResponseHandler}
+ * @param <T> TransportResponse
+ */
+public class TraceableTransportResponseHandler<T extends TransportResponse> implements TransportResponseHandler<T> {
+
+    private final Span span;
+    private final TransportResponseHandler<T> delegate;
+
+    /**
+     * Constructor.
+     * @param span
+     * @param delegate
+     */
+    private TraceableTransportResponseHandler(TransportResponseHandler<T> delegate, Span span) {
+        this.delegate = Objects.requireNonNull(delegate);
+        this.span = Objects.requireNonNull(span);
+    }
+
+    /**
+     * Factory method.
+     * @param delegate delegate
+     * @param span span
+     * @return transportResponseHandler
+     */
+    public static TransportResponseHandler create(TransportResponseHandler delegate, Span span) {
+        if (FeatureFlags.isEnabled(FeatureFlags.TELEMETRY) == true) {
+            return new TraceableTransportResponseHandler(delegate, span);
+        } else {
+            return delegate;
+        }
+    }
+
+    @Override
+    public T read(StreamInput in) throws IOException {
+        return delegate.read(in);
+    }
+
+    @Override
+    public void handleResponse(T response) {
+        try {
+            delegate.handleResponse(response);
+        } finally {
+            span.endSpan();
+        }
+    }
+
+    @Override
+    public void handleException(TransportException exp) {
+        try {
+            delegate.handleException(exp);
```

Comment:
@Gaganjuneja I was looking into different flows how these forest of callbacks is used, I am more and more convinced (unless you have strong arguments not to) that the callback should be called in scope of the current span:

```
try (tracer.withSpanInScope(span)) {
    delegate.handleException(exp);
} finally {
    span.setError(exp);
    span.endSpan();
}
```

Why is that? The callback do non-trivial things, and totally could initiate own spans for which we have to preserve the hierarchy. Check `HandshakeResponseHandler` for example, it delegates to listener(s) that could be also instrumented.

Reply:
It makes sense and there should always be a single flow where span should always be in scope before any operation around it. With this change it would also become the natural flow which now seems to be a special handling. Let me make this change right away.